File: leetcode/22. Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        def dfs(curr, left, right):
            if len(curr) == 2*n:
                res.append(curr)
                return
            if left != n:
                dfs(curr + '(', left+1, right)
            if left > right:
                dfs(curr + ')', left, right+1)
        dfs('', 0, 0)
        return res
            

File: leetcode/22. Generate Parentheses.kt
class Solution {
    fun generateParenthesis(n: Int): List<String> {
        val res = mutableListOf<String>()
        fun generate(left: Int, right: Int, curr: String) {
            if (left == 0 && right == 0) {
                res.add(curr)
                return
            }
            if (left < right) {
                generate(left, right-1, curr + ")")
            }
            if (left > 0) {
                generate(left-1, right, curr + "(")
            }
        }

        generate(n, n, "")
        return res
    }
}
